Integrate Memory Monitor

OhmSpectator opened the pull request:
This PR integrates the previously standalone memory-monitor tool into the EVE root filesystem, along with several enhancements and refinements to ensure seamless operation within the EVE environment.

The memory-monitor tool, which monitors the memory usage of the zedbox process, EVE, and pillar cgroups, has been adapted and integrated into the EVE root filesystem. This integration includes creating necessary Dockerfiles, adding services, implementing scripts, and updating build configurations to ensure the tool operates smoothly within the EVE environment. The tool's configuration management has been improved to prioritize user-provided configuration files, allowing for easier customization and management.

Logging has been enhanced to include configuration values at startup, improving visibility and debugging capabilities. Error handling has been refactored to ensure proper operation and avoid undefined behavior. Additionally, operational improvements have been made to ensure necessary application directories are created for writable operation spaces and to introduce a wait loop that ensures the monitoring thread starts only after the zedbox process is fully operational.

Documentation has been updated to reflect the integration, detailing the startup process, configuration management, and deployment instructions.

Scripts and binaries have been renamed for improved clarity and consistency.

OhmSpectator commented:
Fixed the EOL errors in the README file

OhmSpectator commented:
Fixed the shellcheck errors

eriknordmark commented:
Note that yetus only provides the first 10 errors the way we run it on a PR; seems like there are 10 markdownlint errors which means that there are likely to be more than 10.

christoph-zededa commented on pkg/memory-monitor/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+FROM lfedge/eve-alpine:9fb9b9cbf7d90066a70e4704d04a6fe248ff52bb as memory-monitor-build
```
```suggestion
FROM lfedge/eve-alpine:1f7685f95a475c6bbe682f0b976f12180b6c8726 as memory-monitor-build
```

OhmSpectator replied:
Do we have a reliable way to get this hash?... And why do we use it at all?.. Why not semantic tags?

christoph-zededa replied:
You are hereby invited to join the discussion on https://github.com/lf-edge/eve/pull/3934 ;-)

OhmSpectator replied:
Fixed

OhmSpectator replied:
Fixed

christoph-zededa commented on pkg/memory-monitor/Makefile:
```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CC=gcc
+CFLAGS=-Wall -static -O3
```
```suggestion
CFLAGS=-Wall -static -O2
```
`O3` might cut corners that we're not sure about, especially if the code lacks tests.

OhmSpectator replied:
Ok, I'll change it.

OhmSpectator replied:
Fixed

christoph-zededa commented on pkg/memory-monitor/README.md:
```diff
@@ -0,0 +1,300 @@
+## How to analyze the heap dump
+
+The heap dump (`heap_pillar.out`) is a text file that can be analyzed using the
+`pprof` tool. The `pprof` tool is a part of the Go toolchain, so it is not
```
```suggestion
`go tool pprof` tool. The `go tool pprof` tool is a part of the Go toolchain, so it is not
```
Otherwise people might come up with the idea to use https://github.com/gperftools/gperftools

OhmSpectator replied:
Will fix, thanks

OhmSpectator replied:
Fixed

christoph-zededa commented on pkg/memory-monitor/README.md:
````diff
@@ -0,0 +1,300 @@
+`heap_pillar.out` file and the `zedbox` binary to the same directory on your
+local machine. Then, run the following command:
+```
+$ go tool pprof -http=:8080 /path/to/zedbox /path/to/output/<event_timestamp>/heap_pillar.out
````
```suggestion
$ go tool pprof /path/to/zedbox /path/to/output/<event_timestamp>/heap_pillar.out
```
so it will automatically choose a http port; and it even works if 8080 is already blocked

OhmSpectator replied:
Good to know, thanks!

OhmSpectator replied:
Fixed

christoph-zededa commented on pkg/memory-monitor/memory-monitor-ctr-init.sh:
```diff
@@ -0,0 +1,13 @@
+/sbin/memory-monitor
+
+# Let this script run forever, while the memory monitor is running
+while pgrep "/sbin/memory-monitor" > /dev/null ; do
+    sleep 30
+done
```
Isn't this what `waitpid` is for?

OhmSpectator replied:
Hmm... I'll check. Did not know about the utility.

OhmSpectator replied:
I've removed the entire script

christoph-zededa commented on pkg/memory-monitor/src/monitor/procfs.c:
```diff
@@ -0,0 +1,95 @@
+            // The handler script has already been executed
+            return 0;
+        syslog(LOG_INFO, "----- Zedbox threshold is reached -----\n");
+        char event_msg[256];
```
```suggestion
        char event_msg[103];
```
At least that's what I count ... ;-)

OhmSpectator replied:
We can be generous and give them 153 bytes more) But I can replace it with some meaningful constant if you wish.

OhmSpectator commented:
> Note that yetus only provides the first 10 errors the way we run it on a PR; seems like there are 10 markdownlint errors which means that there are likely to be more than 10.

@eriknordmark, I have not touched the markdown linter error yet. I see more than 10 and will fix )

rouming commented on pkg/memory-monitor/memory-monitor-ctr-init.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
Do we need this starter and daemonization if you can run the whole monitor as a tool from Dockerfile directly?

OhmSpectator replied:
We discussed it with @rucoder, and it looks like it's still necessary. The problem is that the tool is demonized, which means it goes to the background and returns control to the caller. But our boot system (what's the proper name? LinuxKit? containerd?) does not expect a service to return control; that's why we need a wrapper with an infinite loop.

OhmSpectator replied:
There is even a comment about it in the script =)

christoph-zededa replied:
but you can just remove this, can't you? - https://github.com/lf-edge/eve/pull/3984/files#diff-f17e11f9dca74dfea2d7e3b33156609477e2745002fa9f7fc072c0490b746100R85

rouming replied:
That what I'm saying, why you need to daemonize the tool if you wait for it?

OhmSpectator replied:
To clarify, I've added a section about the standalone run into the documentation.

rouming replied:
Sure thing, keep it as you wish, but initial decision should not stop you from implementing the simple flag 
`  -d            Don't run as daemon and log to stderr'` (copied from chronyd).

OhmSpectator replied:
That makes sense. @rucoder also insisted on such a flag. I will add the flag then and retest how the tool behaves in a foreground version. At least I had some problems with the daemonized behaviour when I implemented it (logging, basically, but I can forget something else).

OhmSpectator replied:
Done

OhmSpectator commented:
I hope I've fixed the markdownlint warnings...

OhmSpectator commented:
I think I've taken into account all the comments and issues.

christoph-zededa commented on pkg/memory-monitor/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY --from=memory-monitor-build /root/memory-monitor/dist/memory-monitor /sbin/
+COPY --from=memory-monitor-build /root/memory-monitor/dist/memory-monitor-handler.sh /sbin/
+COPY --from=memory-monitor-build /root/memory-monitor/dist/memory-monitor.conf /etc/
+# COPY --from=memory-monitor-build /root/memory-monitor/dist/sbin.memory-monitor.memory-monitor-handler /etc/apparmor.d/
```
```suggestion
COPY --from=memory-monitor-build /root/memory-monitor/dist/sbin.memory-monitor.memory-monitor-handler /etc/apparmor.d/
```

OhmSpectator replied:
It will not help deliver it to the host `/etc/` directory as it will end in the container version of FS. 
It's still not tested, and I'm already unsure if we need it. Maybe I should just remove this line and integrate the AppArmor profile later.

christoph-zededa replied:
I think it would be good to have a comment here what it is and what is the plan with it.

eriknordmark commented on pkg/memory-monitor/src/monitor/cgroups.c:
```diff
@@ -0,0 +1,332 @@
+    char str_usage[256];
+    unsigned long usage_bytes;
+
+    snprintf(path, sizeof(path), "%s/%s/memory.usage_in_bytes", CGROUP_PATH_PREFIX, cgroup_name);
```
Shouldn't this be sizeof(path)-1 but writing a NUL character to the last byte to ensure that it is always null terminited? Or has snprintf gotten smart enough to do that itself?

OhmSpectator replied:
As far as I know, `sprintf()` always terminates the resulting buffer with `\0`, and it's included in the size argument.

eriknordmark replied:
You're  right - the snprintf() standard manpage says it does.

OhmSpectator commented:
Not sure why the PR builds are failed... =\

OhmSpectator commented:
> Not sure why the PR builds are failed... =\

Actually, I know, I will fix it now.

OhmSpectator commented:
I'm also
* adding a command to the eve tool to reload the config
* adding the output handling to the collect-info.sh script

OhmSpectator commented:
* Added a command to the eve tool to reload the config.
* Added the tool output handling to the collect-info.sh script

OhmSpectator commented:
@uncleDecart , @christoph-zededa is it a "regular" test failure: 
https://github.com/lf-edge/eve/actions/runs/9551860219/job/26328227599#step:3:2779
?..

uncleDecart commented:
@OhmSpectator see https://github.com/lf-edge/eden/pull/969, there was a problem with upgrading tests

milan-zededa commented:
> @uncleDecart , @christoph-zededa is it a "regular" test failure: https://github.com/lf-edge/eve/actions/runs/9551860219/job/26328227599#step:3:2779 ?..

Yes, this one is known. Soon I will have patches for eden ready. For now we have to ignore these failures.

uncleDecart commented on pkg/memory-monitor/src/monitor/cgroups.h:
```diff
@@ -0,0 +1,19 @@
+
+#define CGROUP_PATH_PREFIX "/sys/fs/cgroup/memory"
+#define EVE_CGROUP         "eve"
+#define PILLAR_CGROUP      "eve/services/pillar"
```
Why not constants instead of macros? during debugging building, you'll see string, not variable.

OhmSpectator replied:
And why is it better?

uncleDecart replied:
No magical numbers, constants are handled by compiler so type safety

OhmSpectator replied:
And here are also no magic numbers... We name these constants. And we know their type...

uncleDecart commented on pkg/memory-monitor/src/monitor/cgroups.c:
```diff
@@ -0,0 +1,332 @@
+        return 0;
+    }
+
+    close(fd);
```
Maybe we want to check close error code? This could be potential leak

OhmSpectator replied:
We can check. But what is the reaction in the case of error? Exit entire program? 
```
EBADF  fd isn't a valid open file descriptor.

EINTR  The close() call was interrupted by a signal; see
              signal(7).

EIO    An I/O error occurred.
``|

uncleDecart replied:
I would log the fact that fd was not closed and continued work. I think exiting program would be too dramatic :)

shjala replied:
this is a read, who cares 😄 no cache or anything to be worried about.

uncleDecart commented on pkg/memory-monitor/src/monitor/cgroups.c:
```diff
@@ -0,0 +1,332 @@
+        return 0;
+    }
+
+    // Read the file line by line
```
Are we sure that lines in memory.stat file won't be longer than 256 bytes? Otherwise we will read line in multiple chunks, or can we guarantee that it'll be like "\n\rtotal_cache". And also that the number after will be 256 - strlen("total cache ")? Because we strtoudec will consider NULL as an error

OhmSpectator replied:
Yes, we can guarantee. The format of the file is standard and we know what to expect there.
And about the size:
256 - 12 = 244. Even 240... 10**240 Kb... That's astronomical amount of memory. Yes, we will fit.

uncleDecart reviewed (approved):
LGTM, good job @OhmSpectator. Was nice reviewing C code :D

christoph-zededa commented on pkg/memory-monitor/src/monitor/cgroups.c:
```diff
@@ -0,0 +1,332 @@
+    unsigned long total_cache = 0;
+
+    // Construct the path to the memory.stat file
+    snprintf(buf, sizeof(buf), "%s/%s/memory.stat", CGROUP_PATH_PREFIX, cgroup_name);
```
You give `PATH_MAX` bytes (which is 4k?) in https://github.com/lf-edge/eve/pull/3984/files#diff-5840693cf44b592e67314a102e3c3d3a8a3a6e0fb488d13e832378741cf9b75cR29 but here only 256 bytes for the buffer.

Also you don't check the return value of `snprintf` ...

OhmSpectator replied:
Oops, yeah, I can replace it with PATH_MAX to be cleaner...

OhmSpectator replied:
Fixed. I also updated all the places where I work with char arrays and the snprintf function.

shjala commented on pkg/memory-monitor/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+COPY --from=memory-monitor-build /root/memory-monitor/dist/memory-monitor.conf /etc/
+
+# TODO: Later we will use apparmor to confine the memory-monitor. The apparmor profile template is in dist directory.
+# COPY --from=memory-monitor-build /root/memory-monitor/dist/sbin.memory-monitor-handler /etc/apparmor.d/
```
The apparmor project goes into `pkg/apparmor/profiles`.

OhmSpectator replied:
Ahaaa... Good to know, thanks!

OhmSpectator replied:
Do you think we should add a proper profile with this PR? In this case, I'll try to test it, but there is a chance I'll need some help. Could you review the current template?
https://github.com/OhmSpectator/eve/blob/feature/integrate-memory-monitor/pkg/memory-monitor/src/monitor/sbin.memory-monitor-handler
Also worth noting is that the AppArmor profile exists only for the handler script but not for the monitor itself. Do we need both, or can we somehow combine them?

shjala commented on pkg/memory-monitor/build.yml:
```diff
@@ -0,0 +1,18 @@
+    - /usr/bin:/usr/bin:ro # for the ctr command (it's a dependency of the eve command, bleeugh)
+  pid: host
+  capabilities:
+    - all
```
let's break this down and be more granular.

OhmSpectator replied:
Would be nice if you help with it =)

OhmSpectator replied:
So, am I right that to get the proper list, I have to run the tool with all the caps on via `starce` and then - analyze the trace?... Isn't there a tool that can somehow automate it?

OhmSpectator replied:
Ah, ok, there is `capsh`... I'll try to use it.

OhmSpectator replied:
Nope, it looks like it's not suitable... I'm still trying to get the list of necessary caps.

shjala replied:
while I'm reviewing the code I will write down CAPs you need based on the syscalls.

OhmSpectator replied:
Done. See the last commit I've pushed https://github.com/lf-edge/eve/pull/3984/commits/0bf771ad5d23ef1fa42f876da544115c5d9b135c

OhmSpectator replied:
@shjala, I've done it by disabling all caps and fixing the errors I saw.
But @christoph-zededa proposed another interesting solution!
https://github.com/bpftrace/bpftrace/blob/master/tools/capable.bt
I think it would be nice if we added a tool/doc on how to define a proper list of caps.

shjala commented on pkg/memory-monitor/build.yml:
```diff
@@ -0,0 +1,18 @@
+  devices:
+    - path: all
+      type: a
+  privileged: yes
```
scary 😨.

OhmSpectator replied:
If a security expert says to do so, I'll do so =)
I have to understand how to do it properly...

shjala replied:
sure, let's get that merged and then do it based on the resource the memory-monitor needs.

OhmSpectator replied:
Removed. Works fine...

shjala commented on pkg/memory-monitor/build.yml:
```diff
@@ -0,0 +1,18 @@
+    - /dev:/dev # for syslog (/dev/log)
+    - /run:/run:ro # for the zedbox pid file
+    - /sys/fs/cgroup/memory/:/sys/fs/cgroup/memory/ # to read and modify memory limits
+    - /proc:/proc #:ro to read memory usage in /proc/<PID> and RW for the eve command (???)
```
what we just mount it as `ro`, do we care that all other `eve` commands work in the memory-monitor container?

OhmSpectator replied:
I attempted to mount this as read-only, but the commands in my script didn't work. I can't recall the specifics, but it seemed to me that /proc needed to be read-write for ANY `eve` run.

OhmSpectator replied:
I checked and figured out that RW access was necessary only for the `eve exec` command. I reworked the image, and now I can run the necessary tools from the memory-monitor container without running this command. So, I removed RW access.

OhmSpectator commented on pkg/memory-monitor/Makefile:
```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CC=gcc
+CFLAGS=-Wall -static -O2
```
I like the idea, I'll add it.

OhmSpectator replied:
`-fsanitize=bounds-strict`, as far as I understand, requires libubsan... I cannot find it in the Alpine repos =(
And another question: why have you chosen these specific flags? There are also other options for sanitization and some additional flags related to linking securities.

OhmSpectator replied:
The other flags are okay. At least I was able to build the tool successfully.

shjala replied:
> why have you chosen these specific flags?

no specific reason, I know this one form past experience.

OhmSpectator replied:
Actually, `-ftrivial-auto-var-init=zer` does not work as well =(

OhmSpectator replied:
Ahhh... It works in Alpine 3.20, but not 3.16 that we use =(

OhmSpectator replied:
I've integrated the flags suitable for our environment/Alpine version.

OhmSpectator commented:
Hey @christoph-zededa and @shjala, do you notice any important issues in the PR? I'm working on addressing all your comments, but there's a request to merge it sooner rather than later. If there's something we can address in future PRs, it would be great.
I'm currently adding the security flags and will attempt to remove the `privileged` property from the container image. The AppArmor integration task I would postpone for an upcoming PR. Your opinion?

OhmSpectator commented:
I've also removed the `privileged` option from the container.

shjala commented:
> Hey @christoph-zededa and @shjala, do you notice any important issues in the PR? I'm working on addressing all your comments, but there's a request to merge it sooner rather than later. If there's something we can address in future PRs, it would be great. I'm currently adding the security flags and will attempt to remove the `privileged` property from the container image. The AppArmor integration task I would postpone for an upcoming PR. Your opinion?

Can you wait and merge it at start of next week?

OhmSpectator commented:
> Can you wait and merge it at start of next week?

@shjala, I think so. Do you need extra time to general review it? Or would you like me to look at something specific? I would rather not postpone me working on the code while I remember it =D

OhmSpectator commented:
@shjala will we continue with the review today?

shjala commented on pkg/memory-monitor/src/monitor/main.c:
```diff
@@ -0,0 +1,203 @@
+
+    // Create a new application directory, if it doesn't exist
+    if (access(APP_DIR, F_OK) == -1) {
+        if (mkdir(APP_DIR, 0755) == -1) {
```
`group` and `other` also need execute perm?

OhmSpectator replied:
I'll try to remove it now.

OhmSpectator replied:
Mmm... actually, it's a directory permission. Which means X here is for accessing/listing. And that's expected from a directory with the output and configuration that a user can access it via SSH for example. I can try to remove the W permission from the group and others, but I think X should present.

shjala replied:
my bad, you are right, keep the `x`.

shjala commented:
Lets do the UID/Apparmor adjustment in another PR, except the two comments, LGTM.

OhmSpectator commented on pkg/memory-monitor/src/pressure/pressure.c:
```diff
@@ -0,0 +1,36 @@
+    }
+    printf("Press Enter to release memory:");
+    getchar();
+    for (int i = 0; i < pages_to_allocate; i++) {
```
Yeah... In theory, the system should reclaim the memory back after the process is terminated.
But this `free` does not hurt, does it?

OhmSpectator commented:
> Lets do the UID/Apparmor adjustment in another PR

I've created a ticket for that.
